include/libtorrent/ssl/context.hpp

```diff
@@ -0,0 +1,279 @@
+		BIO_free(bp);
+
+		if (!cert)
+		{
+			X509_STORE_free(cert_store);
+			return ec = error_code(int(ERR_get_error()), error::get_ssl_category());
+		}
+
+		// add cert to cert_store
+		X509_STORE_add_cert(cert_store, cert);
+		X509_free(cert);
+
+		// and lastly, replace the default cert store with ours
+		SSL_CTX_set_cert_store(native_handle(), cert_store);
+		return ec;
+#elif defined TORRENT_USE_GNUTLS
+		std::string cert(static_cast<char const*>(certificate.data()), certificate.size());
+
+        gnutls_datum_t ca;
+        ca.data = reinterpret_cast<unsigned char*>(const_cast<char*>(cert.c_str())); // must be null terminated
+        ca.size = cert.size();
+
+		// Warning: returns the number of certificates processed or a negative error code on error
+		int ret = gnutls_certificate_set_x509_trust_mem(native_handle()
+				, &ca
+				, format == pem ? GNUTLS_X509_FMT_PEM : GNUTLS_X509_FMT_DER);
+		return ec = ret >= 0 ? error_code() : error_code(ret, error::get_ssl_category());
+#endif
+	}
+
+#if defined TORRENT_USE_OPENSSL
+	error_code set_servername_callback(std::function<bool(stream_base& s, std::string name)> cb, error_code& ec)
+	{
+#if OPENSSL_VERSION_NUMBER >= 0x90812f
+#if defined __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wold-style-cast"
+#endif
+		SSL_CTX_set_tlsext_servername_callback(native_handle(), (cb ? servername_func : nullptr));
+		SSL_CTX_set_tlsext_servername_arg(native_handle(), (cb ? m_impl.get() : nullptr));
+#if defined __clang__
+#pragma clang diagnostic pop
+#endif
+		m_impl->servername_callback = std::move(cb);
+		return ec;
+#else // OPENSSL_VERSION_NUMBER
+		return ec = {boost::asio::error::operation_not_supported, error::get_ssl_category()};
+#endif
+	}
+#endif
+
+private:
+	static void global_init();
+	static bool verify_dn(std::string const& expected, bool preverified, base::verify_context& ctx);
+
+#ifdef TORRENT_USE_OPENSSL
+	static int servername_func(SSL* ssl, int*, void *arg);
+#endif
+
+	struct impl
+	{
+		impl(context* p) : parent(p) {}
+		~impl() = default;
+
+		context* parent;
+
+#ifdef TORRENT_USE_OPENSSL
+		std::function<bool(stream_base& s, std::string name)> servername_callback;
+#endif
+	};
+
+	std::shared_ptr<impl> m_impl;
+	friend class stream_base;
+};
+
+#ifdef TORRENT_USE_OPENSSL
+// by openssl changelog at https://www.openssl.org/news/changelog.html
+// Changes between 1.0.2h and 1.1.0  [25 Aug 2016]
+// - Most global cleanup functions are no longer required because they are handled
+//   via auto-deinit. Affected function CRYPTO_cleanup_all_ex_data()
+#if !defined(OPENSSL_API_COMPAT) || OPENSSL_API_COMPAT < 0x10100000L
+namespace {
+	// openssl requires this to clean up internal
+	// structures it allocates
+	struct cleanup
+	{
+#ifdef TORRENT_MACOS_DEPRECATED_LIBCRYPTO
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#endif
+		~cleanup() { CRYPTO_cleanup_all_ex_data(); }
```

Comment:
I would prefer this code also lives in a .cpp file. `session_impl.cpp` is probably *not* the best place, but some counterpart to the previous `openssl.cpp` would make sense I think